### ISSUE TITLE
How to Build TensorFlowLiteGPUExperimental for iOS x86 + others

### ISSUE THREAD

**Issue by airman00** [tensorflow/tensorflow#47978]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): macOS Big Sur
- TensorFlow version: 2.5.0
- Python version:  3.8.3
- Bazel version (if compiling from source):  3.7.2
- GCC/Compiler version (if compiling from source): 12

**Describe the problem**
There is a great pod called 'TensorFlowLiteGpuExperimental' that has C++ support for GPU delegates on iOS devices. Until now I've been using it as a pod and using the C++ API. It's been working great.

The contents of the Pod is a .framework file. Contents of the framework file are attached as screenshots at the end of this post. Most importantly it seems to include the flatbuffers and tensorflow header files, as well as libmetal_delegate.a

I want to build this framework locally so I can add x86 support so I can use this framework in an iOS simulator. Currently it runs great on iOS devices but is lacking x86 support.

I think the solution is writing some custom build instructions for Bazel, but after many many attempts I'm still pretty lost.

I think the solution lies somewhere in this BUILD.apple file
https://github.com/tensorflow/tensorflow/blob/b04e4ecfea73a460ecaf80997dc1ca4a01ceeccd/tensorflow/lite/ios/BUILD.apple#L102

My current understanding is that I need to come up with my own build section here and then run that? Is that right?

My own version of this somehow? `bazel build -c opt --config=ios_fat //tensorflow/lite/ios:TensorFlowLiteC_framework`

**Provide the exact sequence of commands / steps that you executed before running into the problem**
1. I've tried  these to see if it would work out of the box, but unfortunately it's not working for C++ support. 
https://github.com/tensorflow/tensorflow/blob/b04e4ecfea73a460ecaf80997dc1ca4a01ceeccd/tensorflow/lite/ios/BUILD.apple
`bazel build -c opt --config=ios_fat //tensorflow/lite/ios:TensorFlowLiteC_framework
`and then separately also doing
`bazel build -c opt --config=ios_fat //tensorflow/lite/ios:TensorFlowLiteCMetal_framework
`

The good news is I see this header that I use `metal_delegate.h`, but I think I want to expose all the headers like the TensorFlowLiteGPUExperimental does (see screenshots at end).

In my Objective C code on the device itself, I include the headers as follows:
```
#include <tensorflow_lite_gpu/tensorflow/lite/interpreter.h>
#include <tensorflow_lite_gpu/tensorflow/lite/kernels/register.h>
#include <tensorflow_lite_gpu/tensorflow/lite/model.h>
#include <tensorflow_lite_gpu/tensorflow/lite/op_resolver.h>
#include <tensorflow_lite_gpu/tensorflow/lite/string_util.h>
#include <tensorflow_lite_gpu/tensorflow/lite/delegates/gpu/metal_delegate.h>
```

edit:
@thaink 
Here are the contents of the original TensorFlowLite GPU pod that I am attempting to replicate by building
![image](https://user-images.githubusercontent.com/2061485/112009196-8e604e80-8afc-11eb-9896-a5e9b5dfc503.png)
![image](https://user-images.githubusercontent.com/2061485/112009215-928c6c00-8afc-11eb-8abf-6a2c8e4589f6.png)

**Comment by airman00** [comment]
Here's my attempt using glob to just include all header files, but then I get no headers at all in the .framework file that outputs

```
# bazel build -c opt --config=ios_fat //tensorflow/lite/ios:TensorFlowLiteGPU_framework
tflite_ios_framework(
    name = "TensorFlowLiteGPU_framework",
    hdrs = glob([
        "tensorflow/lite/*.h",
        ":c_api.h",
        ":common.h",
        ":xnnpack_delegate.h",
        ":metal_delegate.h",
    ]),
    allowlist_symbols_file = ":allowlist_TensorFlowLiteGPU.txt",
    bundle_name = "TensorFlowLiteGPU",
    minimum_os_version = TFL_MINIMUM_OS_VERSION,
    deps = [
        ":tensorflow_lite_c",
        "//tensorflow/lite/delegates/gpu:metal_delegate",
    
    ],
)
```

**Comment by wangtz** [comment]
Hi Thai,

Do you mind have a look at this iOS related issue?

Thanks,
Tiezhen

**Comment by thaink** [comment]
I think it is equivalent to  "TensorFlowLiteCMetal_framework". @airman00 Can you try that one?

**Comment by airman00** [comment]
Hi @thaink , thanks for the reply

I tried that but it's missing all the headers that I need

Here is the headers from the TensorFlowLiteCMetal_Framework:
![image](https://user-images.githubusercontent.com/2061485/113578377-b7024100-95f0-11eb-9ce6-73e7174a5a6a.png)

You can see all the headers that I'd like to have that are included in the gpu_experimental pod

Any ideas on what to try next?

**Comment by thaink** [comment]
Those headers are for C++. Would you like to use C++ for iOS?
Currently, we only export C API for iOS since it can be imported in swift directly.

**Comment by thaink** [comment]
As above reason, TensorFlowLiteGPUExperimental was deprecated long ago.

Our current recommendation as as in https://www.tensorflow.org/lite/performance/gpu#ios_with_xcode

**Comment by airman00** [comment]
Yes, is there any way to build C++ for iOS? 

Do you also recommend using Swift vs ObjC  on iOS or it should not make a difference in terms of performance?

**Comment by airman00** [comment]
How is this pod built? 
'TensorFlowLiteObjC'

Is there any way to see the exact build commands used to replicate that locally?

 I know how to build TensorFlowLiteC.framework but I want to rebuild the entire pod and add x86 support.

**Comment by airman00** [comment]
I solved my problem in a different way.

I just disabled it from building in the first place on Simulator.

For reference:
1. In your Target's Build Settings delete the entire entry VALID_ARCHS (if it exists) in the User Defined section at the bottom
![image](https://user-images.githubusercontent.com/2061485/114323902-5fcf0580-9af5-11eb-8d36-ea262115f4c6.png)

2. Add this in EXCLUDE ARCHITECTURES in your Project level Build Settings
![image](https://user-images.githubusercontent.com/2061485/114323792-efc07f80-9af4-11eb-84aa-1276fdf8c5a0.png)

2a. If you use Pods, you might need to also do step 2 for the Pod Project level Build Settings
